hushline/premium.py

```diff
@@ -0,0 +1,609 @@
+        )
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with customer ID {subscription.customer}")
+
+
+def handle_subscription_updated(subscription: stripe.Subscription) -> None:
+    # customer.subscription.updated
+
+    # If subscription changes to cancel or unpaid, downgrade user
+    user = db.session.scalars(
+        db.select(User).filter_by(stripe_subscription_id=subscription.id)
+    ).one_or_none()
+    if user:
+        user.stripe_subscription_status = StripeSubscriptionStatusEnum(subscription.status)
+        user.stripe_subscription_cancel_at_period_end = subscription.cancel_at_period_end
+        user.stripe_subscription_current_period_end = datetime.fromtimestamp(
+            subscription.current_period_end
+        )
+        user.stripe_subscription_current_period_start = datetime.fromtimestamp(
+            subscription.current_period_start
+        )
+
+        current_app.logger.info("status is: " + subscription.status)
+        if subscription.status in ["active", "trialing"]:
+            user.tier_id = BUSINESS_TIER
+        else:
+            user.tier_id = FREE_TIER
+
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_subscription_deleted(subscription: stripe.Subscription) -> None:
+    # customer.subscription.deleted
+
+    user = db.session.scalars(
+        db.select(User).filter_by(stripe_subscription_id=subscription.id)
+    ).one_or_none()
+    if user:
+        user.tier_id = FREE_TIER
+        user.stripe_subscription_id = None
+        user.stripe_subscription_status = None
+        user.stripe_subscription_cancel_at_period_end = None
+        user.stripe_subscription_current_period_end = None
+        user.stripe_subscription_current_period_start = None
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_invoice_created(invoice: stripe.Invoice) -> None:
+    # invoice.created
+
+    try:
+        new_invoice = StripeInvoice(invoice)
+        db.session.add(new_invoice)
+        db.session.commit()
+    except ValueError as e:
+        current_app.logger.error(f"Error creating invoice: {e}")
+
+
+def handle_invoice_updated(invoice: stripe.Invoice) -> None:
+    # invoice.updated
+
+    stripe_invoice = db.session.scalars(
+        db.select(StripeInvoice).filter_by(invoice_id=invoice.id)
+    ).one_or_none()
+    if stripe_invoice:
+        stripe_invoice.total = invoice.total
+        stripe_invoice.status = StripeInvoiceStatusEnum(invoice.status)
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find invoice with ID {invoice.id}")
+
+
+async def worker(app: Flask) -> None:
+    current_app.logger.error("Starting worker")
+    with app.app_context():
+        while True:
+            stripe_event = db.session.scalars(
```

Comment:
This query/update/commit has a race condition because it's not done within a transaction block. If events can be processed multiple times, this isn't a problem.

Reply:
Fixed in 438630f4a5efc22204768c950bd78e514ef530cd